core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java

```diff
@@ -0,0 +1,316 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.worker.block;
+
+import alluxio.Configuration;
+import alluxio.PropertyKey;
+import alluxio.StorageTierAssoc;
+import alluxio.WorkerStorageTierAssoc;
+import alluxio.exception.BlockDoesNotExistException;
+import alluxio.exception.ExceptionMessage;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.options.OpenOptions;
+import alluxio.util.network.NetworkAddressUtils;
+import alluxio.worker.block.io.BlockReader;
+import alluxio.worker.block.io.BlockWriter;
+import alluxio.worker.block.io.LocalFileBlockWriter;
+import alluxio.worker.block.meta.UnderFileSystemBlockMeta;
+
+import com.google.common.base.Preconditions;
+import com.google.common.io.Closer;
+import io.netty.buffer.ByteBuf;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.ReadableByteChannel;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * This class implements a {@link BlockReader} to read a block directly from UFS, and optionally
+ * cache the block to Alluxio worker if the whole block it read.
+ */
+@NotThreadSafe
+public final class UnderFileSystemBlockReader implements BlockReader {
+  private static final Logger LOG = LoggerFactory.getLogger(UnderFileSystemBlockReader.class);
+
+  /** An object storing the mapping of tier aliases to ordinals. */
+  private final StorageTierAssoc mStorageTierAssoc = new WorkerStorageTierAssoc();
+
+  /** The file buffer size used to allocate memory from Alluxio storage. */
+  private final long mFileBufferSize;
+  /** The block metadata for the UFS block. */
+  private final UnderFileSystemBlockMeta mBlockMeta;
+  /** If set, do not cache the block. */
+  private final boolean mNoCache;
+  /** The Alluxio block store. It is used to interact with Alluxio. */
+  private final BlockStore mAlluxioBlockStore;
+
+  /** The input stream to read from UFS. */
+  private InputStream mUnderFileSystemInputStream;
+  /** The block writer to write the block to Alluxio. */
+  private BlockWriter mBlockWriter;
```

Comment:
BlockWriter (LocalFileBlockWriter) does not keep any state now though it has position kept implicitly by the RandomAccessFile opened in it. It only provides interface to write data to a file. 

I have moved mBlockWrtierPos to BlockWriter. I kept isBlockCached logic in this reader. if we want to put that in the BlockWriter, then i need to pass block size to the writer. I can do that as well if you prefer.